lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <bsd_limits.h>
+#include <nrf_inbuilt_key.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define NRF_INBUILT_KEY_OP_LS "AT%CMNG=1"
+#define NRF_INBUILT_KEY_OP_RD "AT%CMNG=2"
+#define NRF_INBUILT_KEY_OP_WR "AT%CMNG=0"
+#define NRF_INBUILT_KEY_OP_RM "AT%CMNG=3"
+
+#define NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 (10)
+#define NRF_INBUILT_KEY_CR_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_LF_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_QUOTES_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE (1)
+#define NRF_INBUILT_KEY_COMMA_SIZE (1)
+#define NRF_INBUILT_KEY_INVALID_HANDLE (-1)
+
+static u8_t scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	enum at_cmd_state state;
+
+	int retval =
+		at_cmd_write("AT+CMEE?", response, sizeof(response), &state);
+	if (retval == 0) {
+		char *deactivated = strchr(&response[6], '0');
+
+		if (deactivated != NULL) {
+			return 0;
+		}
+
+		char *activated = strchr(&response[6], '1');
+
+		if (activated != NULL) {
+			return 1;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	char response[10];
+	enum at_cmd_state state;
+
+	int retval = at_cmd_write(at_cmee_strings[(int)enable], response,
+				  sizeof(response), &state);
+
+	if (retval < 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
```

Comment:
There is a lot of duplicated code between the nrf_inbuilt_key_* functions, please see if it is possible to re-use some of it.

Reply:
Moved out the CMEE enablement, which was half of the code in each function. You can take a look if it looks better now.